### DISCUSSION TITLE
🚧`fastn 0.5` syntax

### DISCUSSION THREAD

**Discussion by Arpita-Jaiswal** [fastn-stack/fastn#1993] [General]
The `fastn 0.5` syntax builds on `fastn 0.4`, introducing several key modifications to support complex operations. For those new to `fastn 0.4`, you may want to review the [discussion here](https://github.com/orgs/fastn-stack/discussions/1079) to understand its foundational syntax and implementation.

Lets start with simple grammar:

## Understanding `fastn 0.5` p-script grammar

1. **Grammar with section body**

    A basic section definition can include an optional section type, name, caption, multiple headers, and an optional body:
    
    ```html
    -- <section-type>? <section-name>: <caption>?
    [<header-type>? <header-name>: <header-value>]*
    
    <body>?
    ```
    <details>
      <summary>Example</summary>
      
      ```ftd
      -- person john: John Doe
      age: 10
      gender: male
      
      This is John Doe.
      He is a software developer.
      ```
  
      In this example:
      - `<section-type>`: person
      - `<section-name>`: john
      - `<caption>`: John Doe
      - Header 1:
        - `<header-name>`: age
        - `<header-value>`: 10
      - Header 2:
        - `<header-name>`: gender
        - `<header-value>`: male
      - `<body>`: 
          This is John Doe.
          He is a software developer.
    </details> 

2. **Grammar with subsections**

    Multiple subsections within a main section are enclosed within the section's start and end tags:
    
    ```html
    -- <section-type>? <section-name>: <caption>?
    [<header-type>? <header-name>: <header-value>]*
    
    [<subsection>]*
    
    -- end: <section-name>
    ```

    <details>
      <summary>Example</summary>
      
      ```ftd
      -- team: Development Team
      location: San Francisco
      industry: Technology
      
        -- person: Alice Smith
        age: 28
        role: Frontend Developer
      
        -- person: Bob Jones
        age: 35
        role: Backend Developer
        
      -- end: team
      ```
  
      In this example:
      - Main Section:
        - `<section-name>`: team
        - `<caption>`: Development Team
        - Header 1:
          - `<header-name>`: location
          - `<header-value>`: San Francisco
        - Header 2:
          - `<header-name>`: industry
          - `<header-value>`: Technology
        - Subsection 1:
          - `<section-name>`: person
          - `<caption>`: Alice Smith
          - Header 1:
            - `<header-name>`: age
            - `<header-value>`: 28
          - Header 2:
            - `<header-name>`: role
            - `<header-value>`: Frontend Developer
        - Subsection 2:
          - `<section-name>`: person
          - `<caption>`: Bob Jones
          - Header 1:
            - `<header-name>`: age
            - `<header-value>`: 35
          - Header 2:
            - `<header-name>`: role
            - `<header-value>`: Backend Developer
    </details> 

3. **Comment**
    ```html
    ;; <comment>
    
    -- <section-type>? <section-name>: <caption>? ;; <comment>
    ```

    <details>
      <summary>Example</summary>
      
      ```ftd
      ;; This is a standalone comment explaining the team section
      
      -- team development: Dev Team ;; This section contains team members
      location: Remote
      ```
    </details> 

4. **Module Doc Comment**

    Module comments at the beginning of the file provide meta-information. When placed directly above a section, they become section doc comments.
    
    ```html
    == start of file, ignoring comments and empty lines ==
    [;-; <comment>]*
    
    <anything else>
    ```

    <details>
      <summary>Example</summary>
      
      ```ftd
      ;-; This file contains information about the Development Team.
      ;-; Each member's information is organized under the 'team' section.
      
      -- team development: Dev Team
      location: Remote
      ```
    </details> 

5. **Section Doc Comment**

    Each section can include doc comments to describe its purpose or behavior. It should appear right above the section.
    ```html
    [;-; <comment>]*
    -- <section>
    ```

    <details>
      <summary>Example</summary>
      
      ```ftd
      ;-; Information about the project lead for the team Sarah Lee
      -- person lead: Sarah Lee
      age: 40
      role: Project Manager
      
      Sarah Lee oversees project management and ensures timely delivery.
      ```
    </details>

### Symbol Definitions

`*`: can appear zero or more times.
`?`: can appear zero or once.
`+`: must appear one or more times.
`[]`: encloses list of symbols and tokens
`<>`: token

### Token Definitions:

- `<section-type>`: Alphanumeric characters.
- `<section-name>`: Alphanumeric characters.
- `<header-type>`: Alphanumeric characters.
- `<header-name>`: Alphanumeric characters.
- `<body>`: Any characters.
- `<subsection>`: Follows the same syntax as `<section>`.
- `<comment>`: Any characters.
- `<caption> or <header-value>`: Can range from a string to complex f-script expressions, including function scripts and nested sections (explained further below).

## `<caption> or <header-value>`

These tokens allow complex types. They can include anything from simple strings to f-script function calls or even other nested sections, providing flexibility for various use cases.

### 1. Passing value in p-script:

You can pass any character

```
This is a valid caption or header value. It can take emoticons as well 😊.
```

<details>
  <summary>Example</summary>

```ftd
-- taj-mahal: Taj Mahal is one of the seven wonders. 7️⃣
```
</details>

### 2. Passing variable reference in p-script:

```
$name
```

<details>
  <summary>Example</summary>

```ftd
-- show-name: $name
```
</details>

### 3. Passing function in p-script:
```
$full-name(first = $first-name, last = $first-name)
```

<details>
  <summary>Example</summary>

```ftd
-- show-name: $full-name(first = $first-name, last = $first-name)
```
</details>

### 4. Passing value in f-script:

- Single line value
  ```
  { "Hello  ${name}" }
  ```
  
  <details>
    <summary>Example</summary>
  
  ```ftd
  -- greeting: { "Hello  ${name}" }
  ```
  </details>

- Multiline value
  ```
  {
    $"
      Hello  ${name}!
      
      This is your bio.
    "
  }
  ```
  
  <details>
    <summary>Example</summary>
  
  ```ftd
  -- greeting: {
    $"
      Hello  ${name}!
      
      This is your bio.
    "
  }
  ```
  </details>

- Function value

  ```
  {
    let names = ["Bob", "Alice"]
    let result = ""
    for name in names {
        result = result + " " + name
    }
    result
  }
  ```
  
  <details>
    <summary>Example</summary>
  
  ```ftd
  -- greeting: {
      let names = ["Bob", "Alice"]
      let result = ""
      for name in names {
          result = result + " " + name
      }
      result
  }
  ```
  </details>

To understand better about string checkout [Summarising Position On String](https://github.com/orgs/fastn-stack/discussions/1989)

### 5. Passing sections:

```ftd
{
     -- person: Ankit
     age: 10
     
     Hi I am Ankit.

     -- person: Anita
     age: 20
     
     Hi I am Anita.
}
```

<details>
  <summary>Example</summary>

  ```ftd
  -- show-people: {
       -- person: Ankit
       age: 10
       
       Hi I am Ankit.
  
       -- person: Anita
       age: 20
       
       Hi I am Anita.
  }
  ```
</details>

### 6. Passing `string-literals`:

```ftd
Hello { -- warning: $name }! I am {-- highlight: $role }.
```
```ftd
Hello { 
      -- warning: $name 
      affect: highlight

      -- decorator:
}! I am {
      -- highlight: $role 
}.
```

<details>
  <summary>Example</summary>

  ```ftd
  -- about: Hello { -- warning: $name }! I am {-- highlight: $role }.
  ```
</details>

# Introducing new features in `fastn 0.5`

## Markup or `string-literal` 🚧

The string with UI is known as `string-literal` in `fastn`

```ftd
Hello { -- warning: $name }! I am {-- highlight: $role }
```

## `if`, `for` and `match` 🚧

```ftd
-- if $var:

   -- foo: x
   -- bar:

-- end: if
```

```ftd
-- for: (x, index) in $list

-- foo: $x
-- bar: $index

-- end: for
```

```ftd
-- match: $x

-- case 10: 

    -- show-10:
    -- foo:

-- end: match
```

----

More doc: 
- https://github.com/orgs/fastn-stack/discussions/1949
- https://github.com/orgs/fastn-stack/discussions/1954

**Comment by amitu** [comment]
I have also started writing [grammar.bnf](https://github.com/fastn-stack/fastn/blob/a057659b31728952262a3099cce2c8c5160159f4/v0.5/fastn-p1/grammar.bnf) today, mostly inspired by your [EBNF.md](https://github.com/Arpita-Jaiswal/fastn-0.5-compiler/blob/main/spec/EBNF.md).

I am trying to use the exact tokens I have in my tokens.rs file. Also, I am not trying to make the grammar a very valid grammar yet, maybe it will not work with the bnf grammar. The idea is I am trying to create readable grammar instead of "strictly correct" grammar. 

Maybe we can achieve both readable and correct grammar, but a correct grammar over readable grammar does not really give us anything, given that it is handwritten parser.
